### PR TITLE
Extended Master Secret Extension Support

### LINKED ISSUE #86: Extended Master Secret support [RFC7627]
## Summary

We wish to add support for the "Extended Master Secret" extension as recommended by [RFC7627](https://tools.ietf.org/html/rfc7627)

## Motivation

Some dTLS clients use the "Extended Master Secret" extension in the `Client Hello` message. Per the RFC7627 spec, if the client requests "Extended Master Secret" the server must also specify supporting "Extended Master Secret" in the `Server Hello` message. If the server does not, then it is recommended that the client must close the connection and some clients do in fact do this.

What this means in effect, that some dTLS are unable to communicate with a `pion/dtls` server  without this extension support.

## Describe alternatives you've considered

It would be possible to disable "Extended Master Secret" extension in these clients, but increases security risks as describe in RFC7627.

### LINKED ISSUE #89: DTLS with PSK stuck at handshake progress
Once send packet over DTLS with wrong PSK, the DTLS server will be stuck at handshaking and cannot handle other request anymore event use right PSK.

CipherSuites: TLS_PSK_WITH_AES_128_CCM_8

### PR DESCRIPTION
**Pull request by RyanGordon**
#### Description
This adds support for Extended Master Secret extension per https://tools.ietf.org/html/rfc7627

Additionally, while testing, I found some problems and made some stability improvements to improve error propagation and connection closing during handshakes, and an overall handshake timeout (Fixes #89)

#### Reference issue
Fixes #86 
Fixes #89

### CODE REVIEW

**Comment by RyanGordon** [comment]
_Note:_ This does not work yet. I have attempted to follow spec, but something is awry:

### Client
```
/usr/local/Cellar/openssl@1.1/1.1.1c/bin/openssl s_client -dtls -connect 127.0.0.1:4444 -debug -cert cert.pem -key key.pem
```

### Server
```
PIONS_LOG_TRACE=all go run examples/listen/main.go
Listening
dtls TRACE: 17:30:22.813566 conn.go:321: [handshake] -> HelloVerifyRequest
dtls TRACE: 17:30:22.820772 conn.go:321: [handshake] -> ServerHello
dtls TRACE: 17:30:22.820893 conn.go:321: [handshake] -> TypeCertificate
dtls TRACE: 17:30:22.821212 conn.go:321: [handshake] -> ServerKeyExchange
dtls TRACE: 17:30:22.821301 conn.go:321: [handshake] -> ServerHelloDone
dtls TRACE: 17:30:22.823170 conn.go:433: <- ChangeCipherSpec
dtls DEBUG: 17:30:22.823206 conn.go:391: decrypt failed: decryptPacket: cipher: message authentication failed []```

File: client_handlers.go [hunk 1]
```diff
@@ -76,14 +88,20 @@ func clientHandshakeHandler(c *Conn) error {
...
 					if !ok {
 						return fmt.Errorf("Server responded with SRTP Profile we do not support")
 					}
 					c.state.srtpProtectionProfile = profile
+				case *extensionUseExtendedMasterSecret:
+					c.state.extendedMasterSecret = true
 				}
 			}
+			if c.clientRequireExtendedMasterSecret == false && c.state.extendedMasterSecret == false {
+				return fmt.Errorf("Client required Extended Master Secret extension, but server does not support it")
```
Comment (daenney) [inline client_handlers.go]: Based on the error message here I would've expected a conditional of `c.clientRequireExtendedMasterSecret`, not a comparison to `false`?

Comment (RyanGordon) [reply client_handlers.go]: Whops good catch, thanks! This is now fixed

**Comment by Sean-Der** [comment]
Really good work @RyanGordon this is amazing, so great to see this land. Mind squashing down into one commit?

What do you think of making the `Required Master Secret` more general, then it can apply to clients and servers?

I can help write them, but if you can add a few tests this is a merge from me!

**Comment by Sean-Der** [comment]
Oh also multiple commits are fine if you think they are helpful for debugging/future info! Just want to make sure our git history is auditable/useful for others.

thanks again for adding this!

**Pull request closed without merging**